data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchIndexPartitionCreationSupplier.java

```diff
@@ -0,0 +1,116 @@
+        this.openSearchSourceConfiguration = openSearchSourceConfiguration;
+
+        final Object client = clusterClientFactory.getClient();
+
+        if (client instanceof OpenSearchClient) {
+            this.openSearchClient = (OpenSearchClient) client;
+        } else {
+            throw new IllegalArgumentException(String.format("ClusterClientFactory provided an invalid client object to the index partition creation supplier. " +
+                    "The client must be of type OpenSearchClient. The client passed is of class %s", client.getClass()));
+        }
+
+    }
+
+    @Override
+    public List<PartitionIdentifier> apply(final Map<String, Object> globalStateMap) {
+
+        if (Objects.nonNull(openSearchClient)) {
+            return applyForOpenSearchClient(globalStateMap);
+        }
+
+        return Collections.emptyList();
+    }
+
+    private List<PartitionIdentifier> applyForOpenSearchClient(final Map<String, Object> globalStateMap) {
+        IndicesResponse indicesResponse;
+        try {
+            indicesResponse = openSearchClient.cat().indices();
+        } catch (IOException | OpenSearchException e) {
+            LOG.error("There was an exception when calling /_cat/indices to create new index partitions", e);
+            return Collections.emptyList();
+        }
+
+        return indicesResponse.valueBody().stream()
+                .filter(this::shouldIndexBeProcessed)
+                .map(indexRecord -> PartitionIdentifier.builder().withPartitionKey(indexRecord.index()).build())
+                .collect(Collectors.toList());
+    }
+
+    private boolean shouldIndexBeProcessed(final IndicesRecord indicesRecord) {
+        if (Objects.isNull(indicesRecord.index())) {
+            return false;
+        }
+
+        final IndexParametersConfiguration indexParametersConfiguration = openSearchSourceConfiguration.getIndexParametersConfiguration();
+
+        if (Objects.isNull(indexParametersConfiguration)) {
+            return true;
+        }
+
+        final List<OpenSearchIndex> includedIndices = indexParametersConfiguration.getIncludedIndices();
+        final List<OpenSearchIndex> excludedIndices = indexParametersConfiguration.getExcludedIndices();
+
+        boolean matchesIncludedPattern = includedIndices.isEmpty();
+
+
+        for (final OpenSearchIndex index : includedIndices) {
+            final Matcher matcher = index.getIndexNamePattern().matcher(indicesRecord.index());
+
+            if (matcher.matches()) {
+                matchesIncludedPattern = true;
+                break;
+            }
+        }
+
+        boolean matchesExcludePattern = false;
+
+        for (final OpenSearchIndex index : excludedIndices) {
+            final Matcher matcher = index.getIndexNamePattern().matcher(indicesRecord.index());
+
+            if (matcher.matches()) {
+                matchesExcludePattern = true;
+                break;
+            }
+        }
```

Comment:
You can eliminate code duplication, variable re-assignment and breaks by pull this into a separate function:

```
    pubic boolean doesIndexMatchPattern(final List<OpenSearchIndex> indices, final IndicesRecord indicesRecord) {
        for (final OpenSearchIndex index : excludedIndices) {
            final Matcher matcher = index.getIndexNamePattern().matcher(indicesRecord.index());

            if (matcher.matches()) {
                return true;
            }
        }
        return false;
    }
```

Then you can do something simple like:

```
final boolean matchesIncludedPattern = includedIndices.isEmpty() ? true : doesIndexMatchPattern(includedIndices, indicesRecord);
final boolean matchesExcludePattern = doesIndexMatchPattern(excludedIndices, indicesRecord);
return matchesIncludedPattern && !matchesExcludePattern;
```